### ISSUE TITLE
Extension won't work with the --disable-reading-from-canva flag set on Chromium

### ISSUE THREAD

**Issue by oz42** [FelisCatus/SwitchyOmega#715]
I cannot switch to my proxy profile "remote". When I click the option in the menu, nothing happens so this extension is unusable for me. :-(

**Comment by oz42** [comment]
Proxy SwitchySharp works gread without any problem.

**Comment by FelisCatus** [comment]
Could you please right-click on the extension icon and select "Save error log" in the context menu AFTER reproducing the problem? It would be helpful if you attach the log here.

**Comment by oz42** [comment]
Log is attached.

[OmegaLog_1456836906069.txt](https://github.com/FelisCatus/SwitchyOmega/files/152778/OmegaLog_1456836906069.txt)

**Comment by FelisCatus** [comment]
```
 Error: Failed to execute 'getImageData' on 'CanvasRenderingContext2D': The canvas has been tainted by cross-origin data.
    at Error (native)
    at drawIcon (chrome-extension://padekgcemlokbadohgkifijomclgjgif/js/background.js:72:15)
    at chrome-extension://padekgcemlokbadohgkifijomclgjgif/js/background.js:164:16
```

**Comment by FelisCatus** [comment]
Did you somehow changed the SwitchyOmega source? Or are you using a modified version?

Please ensure that you are using the latest version from [SwitchyOmega releases](https://github.com/FelisCatus/SwitchyOmega/releases).

If re-installing the latest official version doesn't help, then it could be a bug in Chrome that mis-triggered the cross-origin data flag. Please try upgrading your Chrome.

Also, did you install any extension that touches all canvas by adding watermarks or similar stuff? You can disable all extensions except SwitchyOmega and restart Chrome and see if it works for you.

**Comment by oz42** [comment]
I am using the latest official version on Slimjet 8.0.2.0.

I have found a workaround:
In Slimjet's Config, I had to uncheck the checkbox "disable identifying you through HTML5 canvas" [1].
I would like to keep it on, so what domain should I put into the exception list to keep this extension happy?

[1] I have thanslated that myself from DE into EN.

**Comment by oz42** [comment]
Ah, here is more information:
http://www.slimjet.com/en/webhelp/prevent-canvas-fingerprinting.htm

**Comment by FelisCatus** [comment]
The extension has a domain of `chrome-extension://padekgcemlokbadohgkifijomclgjgif/` when running. Does the exception list allow such URLs?

P.S. SwitchyOmega doesn't fingerprint you for sure, but it does need to draw the extension icon on canvas (and pass the image data to Chrome for display) in order to work, due to lack of support for dynamic SVG icons in Chrome Extensions. I believe a lot of extensions with dynamic icons rely on this method and they are unlikely to be changed.

**Comment by oz42** [comment]
I really do trust SwitchyOmega, but this filter works globally. Trust is not the problem. ;-)

But it is the first extension that does not work in Slimjet. My other extensions that I use without problems:
uBlock Origin
Vanilla Cookie Manager
Proxy SwitchySharp

I have tried to put chrome-extension://padekgcemlokbadohgkifijomclgjgif/ into the exception list, but this list seems to accept only somewhat normal URLs.

**Comment by FelisCatus** [comment]
I'm not saying that most extension will break. I'm just saying that most extensions **with dynamic icons** could break. Examples of this include graph and bars in icons (e.g. bandwidth meter), animated icons and pictures. This limitation also applies to extensions that let you customize the icon color such as SwitchyOmega.

Proxy SwitchySharp has bundled a bunch of extension icons in different colors so it does not have to use canvas at all. However, SwitchyOmega lets you choose any color you like in the spectrum for profiles, and at most two of them end up in browser icons. Therefore it is not possible without relying on canvas drawing.

Such techniques hadn't cause any problem until this issues pops up on my Github notifications. Yes, I understand the potential harmfulness of canvas fingerprinting and I wholeheartedly support you leaving it on. It is also kind of Slimjet to put it there as an option that could be easily tweaked. Actually I want to turn this on as well, but Chromium seem to hide this behind a flag.

Well, the bad part is, such blocking often results in many false positives. Apart from SwitchyOmega, any website that wants to process image drawn canvas is also broken. This could include Google Docs for example. That's why they have the exception list.

I'm going to close this issue because I believe SwitchyOmega is a legit use case of canvas image data. It is unfortunate that SO doesn't work if your only intention is to hide from canvas fingerprinting. But I don't think it is reasonable to bundle hundreds of extension icons just because one technique of blocking canvas fingerprinting could break it. I mean, if you are into that kind of reasoning, WebRTC shouldn't be used at all because there's also a WebRTC leak. Guess what? Websites will have to do a non-WebRTC version based on browser plugins (NPAPI) or flash, either of which is more dangerous.

As a user of Slimjet, you should open their issue tracker and tell them that **they should allow putting extensions in the white-list.** Or the browser should ask you to grant permission when you install extensions, just as the way you could grant the extensions cross-origin requests.

Or, the blocking can be implemented as a Chrome Extension. Chrome Extensions can't touch other extensions' background page by default, and explicit permission can be granted for this when necessary, thus solving the problem. However, I don't see any mature and reliable solutions on Chrome Web Store, so...

**tl;dr** Won't fix. SwitchyOmega will continue to rely on canvas image data to work. Slimjet should allow white-listing extensions on this feature. Or you could turn the option off and use a Chrome Extension for the same blocking purpose.

**Comment by sluTTYusr** [comment]
If I understand correctly the startup profile (see #747 ) only works with reading from canvas allowed.
This means I would need to sacrifice privacy for coloured icons? Does not make sense imho.
I don't know about other use cases but I mainly use proxies for privacy reasons.

**Comment by FelisCatus** [comment]
@sluTTYusr No, you don't. SO does not require you to sacrifice your privacy nor does it invade your privacy. It just happened that the way that you attempt to protect privacy broke SwitchyOmega as an unexpected side-effect. FYI, you can prevent canvas fingerprinting by painting random pixels on each canvas which completely changes the checksum without changing the image, and I've seen extensions doing this. You can also block reading data from canvas for all web pages but not for extensions.

It is the `--disable-reading-from-canva` flag that breaks SwitchyOmega. Yes, it helps to protect your privacy. But it does so by breaking everything that relies on reading from canvas no matter whether they may hurt your privacy or not. Please open an issue on chromium and ask them to provide a more polite flag which also protects you from fingerprinting, but does not break valid use cases. Or at least one that allows you to white-list stuff.

SO was created long before canvas fingerprinting has received wide media coverage and relied on reading data from canvas for good reason. You can't just won't let anybody draw extension icons on canvas, just because canvas can be used to track you. If you turn on airplane mode just because Internet can be used to collect your privacy data, then you can't blame an online music player for not working properly. It may be a little bit exaggerated but I do think the same principle applies here.

Also, I would certainly do you a favor if it is trivial to work around this in SwitchyOmega. But it is not.

**Comment by sluTTYusr** [comment]
Thanks for taking the time and elaborate!

**Comment by chrcoluk** [comment]
sadly both extensions linked to are broken in new versions of chrome, they both abandonware.

**Comment by FelisCatus** [comment]
I've been thinking about this problem and I want to propose something. Is it acceptable if I just disable the changing of icon? In that case, most of the features would work, but the icon just stays blue forever and never changes because there's no way to draw the icon without canvas. What do you think about this?

**Comment by FelisCatus** [comment]
I've added some processing that tolerates errors when dynamically drawing the icon. If you have disabled canvas reading, or for whatever reason the icon cannot be drawn, all the remaining features should remain at least usable. In that case, the extension icon will stay blue all the time, but you can mouse over the icon to see the activated profile with details.

The 2.4.0 release is on its way with the enhancement.

**Comment by FelisCatus** [comment]
Done in 2.4.0.
